opentelemetry-api/src/opentelemetry/metrics/__init__.py

```diff
@@ -0,0 +1,346 @@
+to generate metrics.
+"""
+
+
+from abc import ABC, abstractmethod
+from logging import getLogger
+from os import environ
+from threading import RLock
+from typing import Optional, cast
+
+from opentelemetry.environment_variables import OTEL_PYTHON_METER_PROVIDER
+from opentelemetry.metrics.instrument import (
+    Counter,
+    DefaultCounter,
+    DefaultHistogram,
+    DefaultObservableCounter,
+    DefaultObservableGauge,
+    DefaultObservableUpDownCounter,
+    DefaultUpDownCounter,
+    Histogram,
+    ObservableCounter,
+    ObservableGauge,
+    ObservableUpDownCounter,
+    UpDownCounter,
+)
+from opentelemetry.util._providers import _load_provider
+
+_logger = getLogger(__name__)
+
+
+class MeterProvider(ABC):
+    def __init__(self):
+        super().__init__()
+        self._lock = RLock()
+
+    @abstractmethod
+    def get_meter(
+        self,
+        name,
+        version=None,
+        schema_url=None,
+    ) -> "Meter":
+        if name is None or name == "":
+            _logger.warning("Invalid name: %s", name)
+
+
+class _DefaultMeterProvider(MeterProvider):
+    def get_meter(
+        self,
+        name,
+        version=None,
+        schema_url=None,
+    ) -> "Meter":
+        with self._lock:
+            super().get_meter(name, version=version, schema_url=schema_url)
+            # This is done in order to make it possible to store configuration
+            # in the meter provider and make it automatically accessible for
+            # any meter even after it changes.
+            meter = _DefaultMeter(name, version=version, schema_url=schema_url)
+            meter._meter_provider = self  # pylint: disable=protected-access
+            return meter
+
+
+class ProxyMeterProvider(MeterProvider):
+    def get_meter(
+        self,
+        name,
+        version=None,
+        schema_url=None,
+    ) -> "Meter":
+        with self._lock:
+            if _METER_PROVIDER:
+                return _METER_PROVIDER.get_meter(
+                    name, version=version, schema_url=schema_url
+                )
+            return ProxyMeter(name, version=version, schema_url=schema_url)
+
+
+class Meter(ABC):
+    def __init__(self, name, version=None, schema_url=None):
+        super().__init__()
+        self._lock = RLock()
+        self._name = name
+        self._version = version
+        self._schema_url = schema_url
+        self._instrument_names = set()
+        self._meter_provider = None
+
+    @property
+    def name(self):
+        with self._lock:
+            return self._name
+
+    @property
+    def version(self):
+        with self._lock:
+            return self._version
+
+    @property
+    def schema_url(self):
+        with self._lock:
+            return self._schema_url
+
+    def _check_instrument_name(self, name):
+        with self._lock:
+            name = name.lower()
+
+            if name in self._instrument_names:
```

Comment:
This behavior COULD be moved into the SDK (rather than the API).  The API (in absence of SDK) should be as close to 0-overhead as possible.

Reply:
> This behavior COULD be moved into the SDK (rather than the API).

This check could be moved into the SDK but it should not. The API spec specifically requires it [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument):

_Meter implementations MUST return an error when multiple Instruments are registered under the same Meter instance using the same name._

The fact that this requirement is in the API document and not in the SDK one tells me that this is a check that should be in the API. It also makes sense to have it in the API since all implementations must run this check. Having it in the SDK means code duplication for all SDK implementations.

> The API (in absence of SDK) should be as close to 0-overhead as possible.

Having code in the API does not cause any kind of overhead. This check lives in the API, in the base class, so that all subclasses that will be implemented in SDKs have access to it, because they need to perform this check. Even if SDK implementations decide to implement their own check they can do so by just not calling `super`. The code in the API methods is not run until `super` is called, so it does not cause any kind of overhead.

Reply:
Then we need to update the metrics API to be more clear here.  You can log failures, but you should not throw.

[Here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/library-guidelines.md#api-and-minimal-implementation) we call out more information about APIs in OpenTelemetry:

> It is important that values returned from this minimal implementation of API are valid and do not require the caller to perform extra checks (e.g. createSpan() method should not fail and should return a valid non-null Span object). The caller should not need to know and worry about the fact that minimal implementation is in effect. This minimizes the boilerplate and error handling in the instrumented code.

However, we also [state](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/error-handling.md#basic-error-handling-principles):

> API methods MUST NOT throw unhandled exceptions when used incorrectly by end users. The API and SDK SHOULD provide safe defaults for missing or invalid arguments. For instance, a name like empty may be used if the user passes in null as the span name argument during Span construction.

IIUC this design here, the error check + log doesn't really allow for an "empty-instrument" to be returned on this failure.   This is why I suggest it's an SDK concern, but assuming both SDK and API don't have to reimplement the lock + check that this requires, I'd just suggest to give yourself a way to return an empty instrument here.

As another note, we were discussing in the Metrics SiG about (in the future) providing an SDK implementation where these errors are thrown at runtime.   We cannot do that be default because of [this line](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/error-handling.md#basic-error-handling-principles):

> The API or SDK may fail fast and cause the application to fail on initialization, e.g. because of a bad user config or environment, but MUST NOT cause the application to fail later at run time, e.g. due to dynamic config settings received from the Collector.

Since instruments can be looked up, we can't just error out there, however it can be handy when testing/debugging to put some kind of layer in front of the SDK which DOES fail, as log-based failures can be a lot harder to diagnose than a stack trace.  

Whatever makes the most sense for python, go for it.   You can keep the checks + lock in the API, I'd just suggest make sure the SDK has access to the result.

Reply:
I have a proposal for a mechanism that will implement this kind of safety (not raising errors), I'll add it outside of this PR because it may also be relevant to tracing.